charms/worker/k8s/src/charm.py

```diff
@@ -234,12 +237,49 @@ def _create_cos_tokens(self):
     @on_error(
         WaitingStatus("Waiting for enable components"), InvalidResponseError, K8sdConnectionError
     )
-    def _enable_components(self):
+    def _configure_components(self):
         """Enable necessary components for the Kubernetes cluster."""
-        status.add(ops.MaintenanceStatus("Enabling DNS"))
-        self.api_manager.enable_component("dns", True)
+        if self._dns_charm_integrated():
+            status.add(ops.MaintenanceStatus("Disabling DNS"))
+            self.api_manager.configure_component("dns", False)
+            self._configure_dns()
+        else:
+            status.add(ops.MaintenanceStatus("Enabling DNS"))
+            self.api_manager.configure_component("dns", True)
+
         status.add(ops.MaintenanceStatus("Enabling Network"))
-        self.api_manager.enable_component("network", True)
+        self.api_manager.configure_component("network", True)
+
+    def _dns_charm_integrated(self) -> bool:
```

Comment:
Maybe this returns a `DNSSettings` [dataclass](https://docs.python.org/3/library/dataclasses.html)
```suggestion
    def _dns_charm_integrated(self) -> [Optional[DNSSettings]]:
```

Reply:
Oh, I just saw your dataclass suggestion. I've used a dictionary now...